plutus-core/plutus-core/src/Bitwise.hs

```diff
@@ -0,0 +1,569 @@
+        (d, r) -> go d <> [fromIntegral r]
+
+byteStringToInteger :: ByteString -> Integer
+byteStringToInteger bs = case BS.uncons bs of
+  Nothing -> 0
+  -- We have to take some care with representations of exact powers of 256, as
+  -- the two's complement in such a case is the identity function. Therefore, if
+  -- we find an 'unpadded' power, we have to presume that it's positive; if we
+  -- find a leading 0x80, but _something_ else is not a zero byte, we assume
+  -- it's negative instead.
+  Just (w8, bs') ->
+    let len = BS.length bs
+        f x = evalState (foldM (go x) 0 [len - 1, len - 2 .. 0]) 1 in
+      if | isPositivePowerOf2 w8 bs' -> f bs
+         | bit 7 .&. w8 == zeroBits  -> f bs
+         | otherwise                 -> negate . f . twosCompToPositive $ bs
+  where
+    -- This is essentially the opposite to encoding. However, because
+    -- ByteStrings can be indexed from the end in constant time, we don't need
+    -- to use something like a stack: instead, we start from the end, and
+    -- accumulate the radix base as we go, increasing the further along we get.
+    -- This is more efficient, as we'd otherwise first have to compute the
+    -- largest power of 256 we need, then divide down, essentially doing the
+    -- work _twice_.
+    go :: ByteString -> Integer -> Int -> State Integer Integer
+    go bs' acc i = do
+      mult <- get
+      let byte = BS.index bs' i
+      modify (256 *)
+      pure $ acc + (fromIntegral byte * mult)
+
+{-# NOINLINE popCountByteString #-}
+popCountByteString :: ByteString -> Integer
+popCountByteString bs = unsafeDupablePerformIO . unsafeUseAsCStringLen bs $ go
+  where
+    -- We use a standard 'big step, small step' approach. The reason for this is
+    -- that bit counting (via a FiniteBits instance) is defined for much larger
+    -- types than Word8. We can thus read 8-blocks of bytes as 64-bit words
+    -- instead (as we don't care about sign and GHC ensures word alignment),
+    -- which gives us potentially up to an 8x speedup.
+    --
+    -- Thus, our 'big step, small step' approach first walks as much of its
+    -- input as it can using steps whose size is Word64, then finishes the job
+    -- with steps whose size is Word8. We use a rank-2 polymorphic method to
+    -- avoid code duplication, since the only operation we need comes from a
+    -- type class, and is thus agnostic to what we're working over. Step size
+    -- can also be determined via Storable in a similar way.
+    go :: (Ptr CChar, Int) -> IO Integer
+    go (ptr, len) = do
+      let (bigSteps, smallSteps) = len `quotRem` 8
+      let bigPtr :: Ptr Word64 = castPtr ptr
+      let smallPtr :: Ptr Word8 = castPtr . plusPtr ptr $ bigSteps * 8
+      bigCount <- countBits bigPtr bigSteps
+      smallCount <- countBits smallPtr smallSteps
+      pure . fromIntegral $ bigCount + smallCount
+
+-- We use a standard 'big step, small step' construction for all the operators
```

Comment:
I would be happy to provide benchmarks if you would like them. The main reason I don't have any is because this was written to deadline, and I figured having working tested code was a priority.